server/app/services/program/ProgramServiceImpl.java

```diff
@@ -218,14 +218,67 @@ public ErrorAnd<ProgramDefinition, CiviFormError> updateProgramDefinition(
             .join());
   }
 
+  private boolean localizationStatusesAreInSyncWithConfiguredStatuses(
+      LocalizationUpdate localizationUpdate, ProgramDefinition program) {
+    ImmutableList<String> localizationStatusNames =
+        localizationUpdate.statuses().stream()
+            .map(LocalizationUpdate.StatusUpdate::configuredStatusText)
+            .collect(ImmutableList.toImmutableList());
+    ImmutableList<String> configuredStatusNames =
+        program.statusDefinitions().getStatuses().stream()
+            .map(StatusDefinitions.Status::statusText)
+            .collect(ImmutableList.toImmutableList());
+    return localizationStatusNames.equals(configuredStatusNames);
+  }
+
   @Override
   public ErrorAnd<ProgramDefinition, CiviFormError> updateLocalization(
-      long programId, Locale locale, String displayName, String displayDescription)
-      throws ProgramNotFoundException {
+      long programId, Locale locale, LocalizationUpdate localizationUpdate)
+      throws ProgramNotFoundException, OutOfDateStatusesException {
     ProgramDefinition programDefinition = getProgramDefinition(programId);
     ImmutableSet.Builder<CiviFormError> errorsBuilder = ImmutableSet.builder();
-    validateProgramText(errorsBuilder, "display name", displayName);
-    validateProgramText(errorsBuilder, "display description", displayDescription);
+    validateProgramText(errorsBuilder, "display name", localizationUpdate.localizedDisplayName());
+    validateProgramText(
+        errorsBuilder, "display description", localizationUpdate.localizedDisplayDescription());
+
+    // It is permissible to set the localized text / email for a status to empty, which falls back
+    // on the English text.
+    if (!localizationStatusesAreInSyncWithConfiguredStatuses(
+        localizationUpdate, programDefinition)) {
+      throw new OutOfDateStatusesException();
+    }
+
+    // We iterate the existing statuses along with the provided statuses since they were verified
+    // to be consistently ordered above.
+    ImmutableList.Builder<StatusDefinitions.Status> toUpdateStatusesBuilder =
+        ImmutableList.builder();
+    for (int statusIdx = 0;
+        statusIdx < programDefinition.statusDefinitions().getStatuses().size();
+        statusIdx++) {
+      LocalizationUpdate.StatusUpdate statusUpdateData =
+          localizationUpdate.statuses().get(statusIdx);
+      StatusDefinitions.Status existingStatus =
+          programDefinition.statusDefinitions().getStatuses().get(statusIdx);
+      StatusDefinitions.Status.Builder updateBuilder =
+          StatusDefinitions.Status.builder()
```

Comment:
suggestion: If we add [toBuilder()](https://github.com/google/auto/blob/master/value/userguide/builders-howto.md#-initialize-a-builder-to-the-same-property-values-as-an-existing-value-instance) to the AutoValue we can instead replace these two lines with `existinStatus.toBuilder()` which I think is a lot more readable in general.

Reply:
YEESSSSS! Done. I imagine there are a lot more places we could likely do that. I'll hunt them down in a follow-on PR.